client/finality-grandpa/src/environment.rs

```diff
@@ -815,6 +815,59 @@ where
 		Ok(())
 	}
 
+	fn concluded(
+		&self,
+		round: RoundNumber,
+		state: RoundState<Block::Hash, NumberFor<Block>>,
+		_base: (Block::Hash, NumberFor<Block>),
+		historical_votes: &HistoricalVotes<Block>,
+	) -> Result<(), Self::Error> {
+		debug!(
+			target: "afg", "Voter {} concluded round {} in set {}. Estimate = {:?}, Finalized in round = {:?}",
+			self.config.name(),
+			round,
+			self.set_id,
+			state.estimate.as_ref().map(|e| e.1),
+			state.finalized.as_ref().map(|e| e.1),
+		);
+
+		self.update_voter_set_state(|voter_set_state| {
+			// NOTE: we don't use `with_current_round` here, it is possible that
```

Comment:
The reason we can't use `with_current_round` here is because the round being concluded is no longer current (must have been previously completed).